bokeh/models/graphs.py

```diff
@@ -48,24 +48,56 @@ def from_networkx(graph, layout_function, **kwargs):
         .. warning::
             Only two dimensional layouts are currently supported.
 
+        .. warning::
+            Node attributes labeled 'index' and edge attributes labeled 'start' or 'end' are ignored.
+            If you want to convert these attributes, please re-label them to other names.
+
         '''
 
         # inline import to prevent circular imports
         from ..models.renderers import GraphRenderer
         from ..models.graphs import StaticLayoutProvider
 
         # Handles nx 1.x vs 2.x data structure change
-        nodes = list(graph.nodes())
-        edges = list(graph.edges())
-
-        edges_start = [edge[0] for edge in edges]
-        edges_end = [edge[1] for edge in edges]
-
-        node_source = ColumnDataSource(data=dict(index=nodes))
-        edge_source = ColumnDataSource(data=dict(
-            start=edges_start,
-            end=edges_end
-        ))
+        # Convert node attributes
+        node_dict = dict()
+        node_attr_keys = [attr_key for node in list(graph.nodes(data=True))
+                          for attr_key in node[1].keys()]
+        node_attr_keys = list(set(node_attr_keys))
+
+        for attr_key in node_attr_keys:
+            node_dict[attr_key] = [node_attr[attr_key] if attr_key in node_attr.keys() else None
+                                   for _, node_attr
+                                   in graph.nodes(data=True)]
+
+        if 'index' in node_attr_keys:
+            from warnings import warn
+            warn("Converting node attributes labeled 'index' are skipped. "
+                 "If you want to convert these attributes, please re-label with other names.")
+
+        node_dict['index'] = list(graph.nodes())
+
+        # Convert edge attributes
+        edge_dict = dict()
+        edge_attr_keys = [attr_key for edge in list(graph.edges(data=True))
+                          for attr_key in edge[2].keys()]
+        edge_attr_keys = list(set(edge_attr_keys))
+
+        for attr_key in edge_attr_keys:
+            edge_dict[attr_key] = [edge_attr[attr_key] if attr_key in edge_attr.keys() else None
+                                   for _, _, edge_attr
+                                   in graph.edges(data=True)]
+
+        if 'start' in edge_attr_keys or 'end' in edge_attr_keys:
+            from warnings import warn
+            warn("Converting edge attributes labeled 'start' or 'end' are skipped. "
+                 "If you want to convert these attributes, please re-label them with other names.")
+
+        edge_dict['start'] = [x[0] for x in graph.edges(data=True)]
```

Comment:
I do not believe you need `data=True` here. I don't think this matters particularly as it should be performant either way, but I would expect without data to be marginally faster.

Reply:
Thank you for pointing it out.
I fixed it and confirmed that there is no problem with both NetworkX 1 and 2.